lib/si-layer-cache/src/retry_queue.rs

```diff
@@ -0,0 +1,609 @@
+        Self {
+            retry_dir,
+            current_backoff: initial_backoff,
+            next_retry_time: Instant::now(),
+            pending_files: BTreeSet::new(),
+        }
+    }
+
+    /// Record a failed retry attempt and update backoff
+    fn record_failure(&mut self, max_backoff: Duration, multiplier: f64) {
+        self.current_backoff =
+            calculate_next_backoff(self.current_backoff, max_backoff, multiplier);
+        self.next_retry_time = Instant::now() + self.current_backoff;
+    }
+
+    /// Record a successful retry and reset backoff
+    fn record_success(&mut self, initial_backoff: Duration) {
+        self.current_backoff = initial_backoff;
+        self.next_retry_time = Instant::now();
+    }
+
+    /// Check if this queue is ready to retry
+    fn is_ready(&self) -> bool {
+        Instant::now() >= self.next_retry_time && !self.pending_files.is_empty()
+    }
+}
+
+/// Handle for tracking a retry attempt
+#[derive(Debug)]
+pub struct RetryHandle {
+    pub(crate) cache_name: String,
+    pub(crate) filename: OsString,
+}
+
+/// Calculate the next backoff duration using exponential backoff with a maximum cap
+fn calculate_next_backoff(current: Duration, max: Duration, multiplier: f64) -> Duration {
+    std::cmp::min(
+        Duration::from_secs_f64(current.as_secs_f64() * multiplier),
+        max,
+    )
+}
+
+/// Determine if a LayerDbError represents a transient failure that should be retried
+pub fn is_retryable_error(error: &LayerDbError) -> bool {
+    match error {
+        // Direct PG errors are always retryable (connection/pool issues)
+        LayerDbError::Pg(_) | LayerDbError::PgPool(_) => true,
+
+        // For PersisterTaskFailed, only retry if PostgreSQL failed
+        // NATS-only failures are not retried - remote services will fetch from PG on cache miss
+        LayerDbError::PersisterTaskFailed(task_error) => task_error.pg_error.is_some(),
+
+        // All other errors are non-retryable
+        _ => false,
+    }
+}
+
+/// Messages sent to RetryQueueManager for queue operations
+#[derive(Debug)]
+pub enum RetryQueueMessage {
+    /// Enqueue a failed event for retry
+    Enqueue(LayeredEvent),
+    /// Mark a retry attempt as successful (removes from queue)
+    MarkSuccess(RetryHandle),
+    /// Mark a retry attempt as failed with retryable error (updates backoff)
+    MarkRetryableFailure(RetryHandle, LayerDbError),
+    /// Mark a retry attempt as permanently failed (removes from queue)
+    MarkPermanentFailure(RetryHandle),
+}
+
+/// Write a LayeredEvent to a retry queue file
+async fn write_retry_file(path: &std::path::Path, event: &LayeredEvent) -> LayerDbResult<()> {
+    use tokio::fs;
+
+    let bytes = postcard::to_stdvec(event)?;
+
+    // Write to temp file first, then atomic rename
+    let temp_path = path.with_extension("tmp");
+    fs::write(&temp_path, bytes)
+        .await
+        .map_err(LayerDbError::RetryQueueFileWrite)?;
```

Comment:
My only thought here is that we might want to call an `fsync` equivalent to prevent the loss of a retry queue file if the system fails before the page buffers have been flushed to disk. (there's `sync_all`, but you need a `File` handle for that). But that might be extra paranoid (and will impact performance)